source/common/quic/quic_stats_gatherer.cc

```diff
@@ -0,0 +1,35 @@
+#include "source/common/quic/quic_stats_gatherer.h"
+
+#include <cstdint>
+
+namespace Envoy {
+namespace Quic {
+
+void QuicStatsGatherer::OnPacketAcked(int acked_bytes,
+                                      quic::QuicTime::Delta /* delta_largest_observed */) {
+  bytes_outstanding_ -= acked_bytes;
+  if (bytes_outstanding_ == 0 && fin_sent_ && !logging_done_) {
+    doDeferredLog();
+  }
+}
+
+void QuicStatsGatherer::doDeferredLog() {
+  logging_done_ = true;
+  for (const std::shared_ptr<StreamInfo::StreamInfo>& stream_info : stream_info_) {
+    if (!stream_info->deferredLoggingInfo().has_value()) {
+      continue;
```

Comment:
Please add a test to verify that in redirect case, the response time is correctly calculated.

Reply:
Done.